dcap-ql-api/src/quote.rs

```diff
@@ -150,6 +158,31 @@ impl<'a> Quote<'a> {
             signature: quote.into(),
         })
     }
+
+    pub fn verify(&self, quote: &[u8]) -> Result<()> {
```

Comment:
In general, signature verification should be a `fn f(signed_data, root_of_trust) -> Result<VerifiedData>`. This ensures that unverified data can't be used. The current APIs don't really fit this flow. I propose the following.

```rust
trait<'a> Quote3SignatureVerify: Quote3Signature<'a> {
    type TrustRoot;

    fn verify(&self, quote: &'a [u8], root_of_trust: &mut Self::TrustRoot) -> Result<()>;
}

trait Quote3SignatureEcdsaP256Verifier {
    /// Verify the platform certification data.
    ///
    /// The certification data is in `quote3signature.certification_data()`.
    ///
    /// On success, should return the platform certification public key (PCK).
    // TODO: pass a container for the certification type and data instead of
    // the whole signature structure.
    fn verify_certification_data(&mut self, quote3signature: &Quote3SignatureEcdsaP256) -> Result<&[u8]>;

    /// Verify the quoting enclave (QE3).
    fn verify_qe3(&mut self, qe3_report: &[u8], authentication_data: &[u8]) -> Result<()>;
}

impl Quote3SignatureVerify for Quote3SignatureEcdsaP256 {
    type TrustRoot = dyn Quote3SignatureEcdsaP256Verifier;

    fn verify(&self, quote: &'a [u8], root_of_trust: &mut dyn Quote3SignatureEcdsaP256Verifier) -> Result<()> {
        let pck = verifier.verify_certification_data(self)?;

        // TODO: verify qe3_signature against qe3_report/authentication_data using pck
        
        verifier.verify_qe3(self.qe3_report(), self.authentication_data())?;
        
        // TODO: verify signature against quote using attestation_public_key
    }
}

impl Quote3 {
    pub fn verify<T: Quote3SignatureVerify>(quote: &'a [u8], root_of_trust: &mut T::TrustRoot) -> Result<Self> {
        let parsed_quote = Self::parse(quote);
        let sig = parsed_quote.signature::<T>()?;
        sig.verify(quote, root_of_trust)?;
        Ok(parsed_quote)
    }
}
```

In the pcs crate, implement Quote3SignatureEcdsaP256Verifier on a stucture holding the QE3 identify, platform cert, root CA etc.